src/ordering/db/buy_order.py

```diff
@@ -0,0 +1,129 @@
+    id: UUID = sa.Column("order_id", sa_utils.UUIDType, unique=True)
+    request_id: UUID = sa.Column(sa_utils.UUIDType, unique=True)
+
+    paid: Decimal = sa.Column(FiatAmountColumn)
+    bought: Decimal = sa.Column(BtcAmountColumn)
+
+    _currency: Currency = sa.Column(
+        "currency", sa.Enum(Currency, naive_enum=False)
+    )
+    _price: Decimal = sa.Column("price", FiatAmountColumn)
+    _rate_date: datetime = sa.Column("rate_date", sa.DateTime)
+
+    @hybrid_property
+    def exchange_rate(self) -> BTCRate:
+        return BTCRate(
+            price=self._price,
+            currency=self._currency,
+            on_date=self._rate_date,
+        )
+
+    @exchange_rate.setter
+    def exchange_rate(self, value: BTCRate) -> None:
+        self._price = value.price
+        self._currency = value.currency
+        self._rate_date = value.on_date
+
+    when_created: datetime = sa.Column(
+        "when_created", sa.DateTime, default=datetime.utcnow, index=True
+    )
+    _when_updated: datetime = sa.Column(
+        "when_updated", sa.DateTime, onupdate=datetime.utcnow,
+    )
+
+    def __repr__(self) -> Text:
+        return (
+            "<"
+            f"{self.__class__.__name__}"
+            f" db_id={self._db_id}"
+            f" order_id={self.id}"
+            f" request_id={self.request_id}"
+            f" amount={self.bought!s}BTC"
+            ">"
+        )
+
+
+@inject
+class ORMRepository(Repository):
+    _session: Session
+    _pending_events: List[Event]
+
+    def __init__(self, transaction: Transaction, bus: EventBus) -> None:
+        self._transaction = transaction
+        self._bus = bus
+
+    @contextmanager
+    def lock(self) -> ContextManager[Decimal]:
+        events = []
+        with self._transaction() as session:
+            self._session = session
+            self._pending_events = events
+
+            lock = f"LOCK TABLE {DBBuyOrder.__tablename__} IN EXCLUSIVE MODE;"
+            session.execute(lock)
+
+            yield self._calculate_balance()
+
+            del self._session
+            del self._pending_events
+
+        for event in events:
+            self._bus.emit(event)
+
+    def _calculate_balance(self) -> Decimal:
+        balance, = self._session.query(func.sum(DBBuyOrder.bought)).one()
+        return Decimal(balance or 0)
+
+    def create(
+            self,
+            request_id: UUID,
+            paid: Decimal,
+            bought: Decimal,
+            with_rate: BTCRate,
+    ) -> BuyOrder:
+        entry = DBBuyOrder()
+        entry.id = uuid4()
+        entry.request_id = request_id
+        entry.paid = paid
+        entry.bought = bought
+        entry.exchange_rate = with_rate
+        self._session.add(entry)
+        self._session.flush()
+        return entry
+
+    def emit(self, event: Event) -> None:
+        self._pending_events.append(event)
+
+    def get_order_id(self, for_request_id: UUID) -> Optional[BuyOrder]:
```

Comment:
Now that you like 3.10 annotations you could also do `BuyOrder | None` I think

Reply:
Yes, I prefer 3.10 annotations. But I really like seeing "**_Optional BuyOrder_**". 
For me, it's more readable than "**_BuyOrder or None_**". But I will change it to 3.10 annotation.